lib/imgbundler/imgbundler.go

```diff
@@ -0,0 +1,136 @@
+package imgbundler
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"oss.terrastruct.com/d2/lib/xmain"
+)
+
+var imageRe = regexp.MustCompile(`<image href="([^"]+)"`)
+
+type resp struct {
+	srctxt string
+	data   string
+	err    error
+}
+
+func InlineLocal(ms *xmain.State, in []byte) ([]byte, error) {
+	return inline(ms, in, false)
+}
+
+func InlineRemote(ms *xmain.State, in []byte) ([]byte, error) {
+	return inline(ms, in, true)
+}
+
+func inline(ms *xmain.State, in []byte, isRemote bool) ([]byte, error) {
+	svg := string(in)
+
+	imgs := imageRe.FindAllStringSubmatch(svg, -1)
```

Comment:
Why not use https://pkg.go.dev/regexp#Regexp.FindAllSubmatch and avoid the extra alloc?

Reply:
what's the extra alloc? it looks like it just gives it in a different form ([]bytes), one that I don't want.

Reply:
Converting from `[]byte` to `string` is an extra alloc. Why can't you just use `[]byte` though?

Reply:
cuz in the next loop i url.Parse the string?

Reply:
Ah right. my b

Reply:
actually in that case add a `string()` in `url.Parse` and then use `bytes` everywhere else. that way it'll be more clear.

Reply:
these are urls. why would it be more clear as bytes?

Reply:
just because something is utf-8 doesn't mean you immediately turn it into a `string`. there's a real cost. you want to keep things as `[]byte` if you can.

here if you only converted to string for `url.Parse`, it'd be clear to me why you're doing that instead of what you're doing now where you're converting the svg all at once at the beginning.

Reply:
it's not even an svg i'm converting. it's literally an image URL. utf-8 has nothing to do with this. what is this real cost other than style here

Reply:
how is line 36 not converting the entire svg? i'm asking you to convert just the image URL instead on `url.Parse`.

Reply:
ohh i didn't see that FindAllSubmatch accepted a []byte too, only that it returned it in that form. okay, agreed

Reply:
actually, i still don't. it's just too many string() conversions. not just in url.Parse, but fetch and read take in string too.

Reply:
Ah, convert the match then and pass around a `urlStr` instead.

Reply:
it's also not convenient (filtering step).

Reply:
Just keep filtered a `[][]string` and do `[]string{string(img[0]), string(img[1])}` when appending. This way you don't need conversions anywhere else other than this loop.